File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    dungeons = list(permutations(dungeons, len(dungeons)))
    
    answer = 0
    for dungeon in dungeons:
        current_k = k
        result = 0
        for item in dungeon:
            need_energy, consume_energy = item
            if current_k >= need_energy:
                current_k -= consume_energy
                result += 1
        answer = max(answer, result)
        
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    let answer = 0;
    const n = dungeons.length;
    const visited = Array(n).fill(false);
    
    const dfs = (count, curr) => {
        answer = Math.max(answer, count);
        
        for (let i = 0; i < n; i++) {
            const [need, use] = dungeons[i];
            if (!visited[i] && curr >= need) {
                visited[i] = true;
                dfs(count + 1, curr - use);
                visited[i] = false;
            }
            
        }
    }
    
    dfs(0, k);
    
    return answer;
}